dataflow/src/main/java/org/checkerframework/dataflow/busyexpression/BusyExprStore.java

```diff
@@ -0,0 +1,152 @@
+import org.checkerframework.dataflow.analysis.Store;
+import org.checkerframework.dataflow.cfg.node.BinaryOperationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.visualize.CFGVisualizer;
+import org.checkerframework.dataflow.expression.JavaExpression;
+import org.checkerframework.javacutil.BugInCF;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.StringJoiner;
+
+/** A busy expression store contains a set of busy expressions represented by nodes. */
+public class BusyExprStore implements Store<BusyExprStore> {
+
+    /** A set of busy expression abstract values. */
+    private final Set<BusyExprValue> busyExprValueSet;
+
+    /**
+     * Create a new BusyExprStore.
+     *
+     * @param busyExprValueSet a set of busy expression abstract values
+     */
+    public BusyExprStore(Set<BusyExprValue> busyExprValueSet) {
+        this.busyExprValueSet = busyExprValueSet;
+    }
+
+    /** Create a new BusyExprStore. */
+    public BusyExprStore() {
+        busyExprValueSet = new LinkedHashSet<>();
+    }
+
+    /**
+     * Kill expressions if they contain variable var.
+     *
+     * @param var a variable
+     */
+    public void killBusyExpr(Node var) {
+        Iterator<BusyExprValue> iter = busyExprValueSet.iterator();
+        while (iter.hasNext()) {
+            BusyExprValue busyExprValue = iter.next();
+            Node expr = busyExprValue.busyExpression;
+            if (isExprContainVariable(expr, var)) {
+                iter.remove();
+            }
+        }
+    }
+
+    /**
+     * Return if the expression contains variable var.
+     *
+     * @param expr the expression checked
+     * @param var the variable
+     * @return true if the expression contains the variable
+     */
+    public boolean isExprContainVariable(Node expr, Node var) {
+        if (expr instanceof BinaryOperationNode) {
+            BinaryOperationNode binaryNode = (BinaryOperationNode) expr;
+            return isExprContainVariable(binaryNode.getLeftOperand(), var)
+                    || isExprContainVariable(binaryNode.getRightOperand(), var);
+        }
+
+        return expr.equals(var);
+    }
+
+    /**
+     * Add busy expression e to busy expression value set.
+     *
+     * @param e the busy expression to be added
+     */
+    public void putBusyExpr(BusyExprValue e) {
+        busyExprValueSet.add(e);
+    }
+
+    /**
+     * Add expressions to the store, add sub-expressions to the store recursively
+     *
+     * @param e the expression to be added
+     */
+    public void addUseInExpression(Node e) {
+        if (e instanceof BinaryOperationNode) {
+            BusyExprValue busyExprValue = new BusyExprValue(e);
+            putBusyExpr(busyExprValue);
+            // recursively add expressions
+            BinaryOperationNode binaryNode = (BinaryOperationNode) busyExprValue.busyExpression;
```

Comment:
Isn't `binaryNode == e`? Should we just use `e`?

Reply:
Though we have known `e` is a instance of `BinaryOperationNode`, we still have to cast it, otherwise compiler will issue an error

Reply:
I think we can write it like this
```
BinaryOperationNode binaryNode = (BinaryOperationNode) e;
```
But I think we still need the type casting.